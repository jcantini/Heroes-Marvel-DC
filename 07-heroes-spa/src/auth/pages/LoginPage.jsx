import { useContext } from "react";
import { useNavigate } from "react-router-dom";
import { AuthContext } from "../context/AuthContext";


export const LoginPage = () => {

  const { login } = useContext( AuthContext ); // Accedo al contexto para tomar la función de login
  const navigate = useNavigate();

  // Cuando se dispara el onLogin, este debe disparar la action de login
  const onLogin = () => {

    login( 'Jorge Cantini' ); // Hago el login del usuario con la  función del provider

    navigate( '/', { // Al dar click en el logout me lleva al home
      replace: true,    // evita que se pueda volver al historial anterior ya que se dio logout
    }); 
}

  return (
    <div className="container mt-5">
      <h1>Login</h1>
      <hr />

      <button 
        className="btn btn-primary"
        onClick={ onLogin}
      >
        Login
      </button>
    </div>

  )
}

