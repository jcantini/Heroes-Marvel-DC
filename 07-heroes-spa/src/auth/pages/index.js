// Sirve para exportar todo lo que tenggo dentro de este directorio asi no tengo
// que hacer tantos imports en AppRouter. Es m√°s prolijo

export * from './LoginPage';