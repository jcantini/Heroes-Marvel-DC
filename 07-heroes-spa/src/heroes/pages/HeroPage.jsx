import { useMemo } from "react";
import { Navigate, useNavigate, useParams } from "react-router-dom"
import { getHeroById } from "../helpers";

export const HeroPage = () => {

  const { id } = useParams(); // (1)

  const navigate = useNavigate();

  // Cada vez que se renderize este componente se vuelve a ejecutar esta función. En este caso no hay 
  // necesidad de volver a renderizar porque solo es para mostrar datos.
  //  De todas formas lo hago para mostrar cómo se hace. 

  // const hero = getHeroById( id ); Esta es la línea que voy a reemplazar
  // Solo se dispara cuando sus dependencias cambien. En este caso cuando cambie el id
  const hero = useMemo( () => getHeroById( id ), [ id ]);  

  const onNavigateBack = () => {
    navigate( -1 ); // -1 lo manda a la página anterior. También puedo detectar si es hero de DC vuelvo a DC
  }                 // si no vuelvo a Marvel

  if( !hero ) {
    return <Navigate to="/marvel"/>
  }

  return (
   <div className="row mt-5">
      <div className="col-4">
        <img 
          src={`/heroes/${id}.jpg`} 
          alt={ hero.superhero }
          className="img-thumbnail animate__animated animate__backInLeft"
        />
      </div>

      <div className="col-8">
        <h3>{hero.superhero}</h3>
        <ul className="list-group list-group-flush">
          <li className="list-group-item"> <b>Alter ego:</b> { hero.alter_ego } </li>
          <li className="list-group-item"> <b>Publisher:</b> { hero.publisher } </li>
          <li className="list-group-item"> <b>First appearance:</b> { hero.first_appearance } </li>
        </ul>

        <h5 className="mt-3">  Characters </h5>
        <p>{ hero.characters }</p>

        <button 
          className="btn btn-primary"
          onClick={ onNavigateBack }  
        >
          Return
        </button>
      </div>
   </div>


  )
}


// (1)
// useParams es un custom hook que lo provee React Router que nos sirve para obtener los parametros
// me devuelve un objeto del cual desestructuro el id que puse cuando defini la ruta.


