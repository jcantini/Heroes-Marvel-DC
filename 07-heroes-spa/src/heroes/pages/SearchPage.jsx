
import { useLocation, useNavigate } from 'react-router-dom';
import queryString from 'query-string';

import { useForm } from '../../hooks/useForm';
import { HeroeCard } from '../components';
import { getHeroesByName } from '../helpers';

export const SearchPage = () => {

  const navigate = useNavigate();
  const location = useLocation(); // (1)
 // console.log({location});  para ver que la propiedad search tiene el string con el query
 // const query = queryString.parse( location.search ); // queryString extrae todo lo que viene en el query
 // console.log( {query} ); // Para ver como lo extrae. Tener en cuenta que siempre devuelve strings
 // Directamente puedo desestructurar la q que es el 1er parámetro y el único en este caso
  const { q = ''} = queryString.parse( location.search ); //q pr default es '' por si no viene nada

  const heroes = getHeroesByName( q ); // Aca tengo el array de heroes encontrados o []

  const showSearch = (q.length === 0); // Devuelve true o false
  const showError = (q.length > 0) && heroes.length === 0;  // hay algo escrito y con eso no se encontro

  const { searchText, onInputChange } = useForm({ // Lo que desestructuro de lo que me devuelve el hook
      searchText: q // inicialmente q='' pero despues ya tiene un valor. Asignando q no lo pierdo si hago refresh
  });


  const onSearchSubmit = ( e ) => {
    e.preventDefault();
  //  if (searchText.trim().length <= 1) return;

    // Quiero armar una ruta de navegación pasándole un query con lo que se ingreso por el input. 
    // Para a navegar a la misma ruta en que me encuentro seria: navigate('')
    // Para navegar agregando un query parameter hago:
    navigate(`?q=${ searchText }`);

  }

  return (
    <>
      <h1>Search</h1>
      <hr />

      <div className="row">
          <div className="col-5">
            <h4>Searching</h4>
            <hr />
            <form aria-label="form" onSubmit={ onSearchSubmit }>
              <input 
                type="text" 
                placeholder="Search a Hero"
                className="form-control"
                name="searchText"
                autoComplete="off"
                value={ searchText }
                onChange={ onInputChange }
              />

              <button className="btn btn-outline-primary mt-2">
                  Search
              </button>
            </form>
          </div>

          <div className="col-7">
            <h4>Results</h4>
            <hr />

            {/* Esto se muestra mientra no se ingreso el texto de search */}
            {/* {
              ( q === '')
              ? <div className="alert alert-primary">Search a hero</div>
              : ( heroes.lenght === 0 )
            } */}
 
            <div aria-label="Found"className="alert alert-primary animate__animated animate__fadeInUp" 
                  style={{ display: showSearch ? '' : 'none'}}>
              Search a hero
            </div>

            {/* Esto se muestra cuando no se encontro */}
            <div aria-label="notFound" className= "alert alert-danger animate__animated animate__fadeInUp" 
                  style={{ display: showError ? '' : 'none'}}
            >
              Ho hero found with name <b>{ q }</b>
            </div>

            {
              heroes.map( hero => (
                  <HeroeCard key= {hero.id} { ...hero } />
              ))
            }


          </div>

      </div>


 
    </>
  )
}

// (1)
// De alguna forma tengo que leer el query parameter. En el navigate no tengo nada y tampoco en params
// Lo obtengo de la localización que me da la ruta donde me encuentro y en esta ruta figura el query