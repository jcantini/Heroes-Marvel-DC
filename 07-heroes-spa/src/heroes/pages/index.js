// Sirve para exportar todo lo que tenggo dentro de este directorio asi no tengo
// que hacer tantos imports en AppRouter. Es m√°s prolijo

export * from './DcPage';
export * from './MarvelPage';
export * from './SearchPage';
export * from './HeroPage';