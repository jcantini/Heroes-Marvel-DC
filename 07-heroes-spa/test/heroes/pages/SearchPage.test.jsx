import { fireEvent, render, screen } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import { SearchPage } from '../../../src/heroes/pages/SearchPage'

const mockedUseNavigate = jest.fn(); // Ver explicaion en Navbar.test
jest.mock('react-router-dom', () => ({ 
    ...jest.requireActual('react-router-dom'),  
    useNavigate: () => mockedUseNavigate 
}));

describe('Pruebas en <SearchPage/>', () => {

    beforeEach( () => jest.clearAllMocks() );

    test('Debe mostrarse correctamente con los valores por defecto', () => {
      // En este caso hago la prueba usando el snapshot. Viendo el componente, se usa el useNavigate
      // y el useLocation por lo que necesito usar el MemoryRouter
      const { container } = render( // tomo el snapshot que guardo en el container
        <MemoryRouter>
            <SearchPage/>
        </MemoryRouter>
      )

      // screen.debug(); // Para ver lo que tengo
      expect( container ).toMatchSnapshot();
    });

    test('Debe mostrar a Batman y el input con el valor del queryString', () => {
        render( // tomo el snapshot que guardo en el container
          <MemoryRouter initialEntries={['/search?q=batman']}>
              <SearchPage/>
          </MemoryRouter>
        )
  
        //screen.debug(); // Para ver lo que 
        
        const input = screen.getByRole('textbox') // capturo el input
        expect( input.value ).toBe('batman');

        // Voy a chequear que se muestre una imagen de batman
        const img = screen.getByRole('img');
        expect(img.src).toContain('/heroes/dc-batman.jpg')  // src lo saque del screen.debug

        // Evaluo que el div que muestra el error tiene display: none
        screen.debug();
        const div = screen.getByLabelText('notFound');
        console.log(div.style) // veo como es la estructura
        expect(div.style.display).toBe('none')
      });

      test('Debe mostrar un error si no se encuenta el hero batman123', () => {
        render( 
          <MemoryRouter initialEntries={['/search?q=batman123']}>
              <SearchPage/>
          </MemoryRouter>
        );
        screen.debug();
        const div = screen.getByLabelText('notFound');
        expect(div.style.display).toBe('')
      });

      test('Debe llamar el navigate a la pantalla nueva', () => {

        const inputValue = 'superman'
        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage/>
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox'); // capturo e input
        fireEvent.change( input, {target: { name: 'searchText', value: inputValue}}) // lo cargo con un valor
        console.log(input.value); // Para comprobar que tengo superman en el input

        // Busco ahora el form al que le tengo que agregar un aria/label xq si no no lo encuentra
        const form = screen.getByRole('form'); // ya accedo al form
        fireEvent.submit( form ); // ejecuto el submit fel form
        
        expect( mockedUseNavigate ).toHaveBeenCalledWith(`?q=${inputValue}`);
    });
      
    
  
})
