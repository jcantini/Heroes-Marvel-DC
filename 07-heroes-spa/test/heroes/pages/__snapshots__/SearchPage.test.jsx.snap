// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage/> Debe mostrarse correctamente con los valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a Hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-2"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        aria-label="Found"
        class="alert alert-primary animate__animated animate__fadeInUp"
      >
        Search a hero
      </div>
      <div
        aria-label="notFound"
        class="alert alert-danger animate__animated animate__fadeInUp"
        style="display: none;"
      >
        Ho hero found with name 
        <b />
      </div>
    </div>
  </div>
</div>
`;
